hxmath/math/DualQuaternion.hx

```diff
@@ -0,0 +1,644 @@
+        return new DualQuaternion(this.real*oneOver, this.dual*oneOver);
+    }
+    
+    /**
+     * Normalize the quaternion in-place.
+     * 
+     * @return  The modified object.
+     */
+    public inline function normalize(): DualQuaternion
+    {
+        var mag = dot( this, this );
+        if( mag < 0 ) return null;
+        var oneOver = 1/mag;
+        var self: DualQuaternion = this;
+        self.real = this.real*oneOver;
+        self.dual = this.dual*oneOver;
+        return self;
+    }
+    
+    private inline function get_length():Float
+    {
+        return Math.sqrt(dot(this, this));
+    }
+    
+    private inline function get_lengthSq():Float
+    {
+        return dot( this, this );
+    }
+    
+    private inline function get_matrix():Matrix4x4
+    {
+        var self: DualQuaternion = this;
+        var m: Matrix4x4 = self;
+        return m;
+    }
+    
+    @:to
+    public function toMatrix4x4(): Matrix4x4
+    {
+        var q = normalize();
+        var m = Matrix4x4.identity;
+        var s = q.real.s;
+        var x = q.real.x;
+        var y = q.real.y;
+        var z = q.real.z;
+        // Extract rotational information
+        m.m00 = s*s + x*x - y*y - z*z;
+        m.m01 = 2*x*y + 2*s*z;
+        m.m02 = 2*x*z - 2*s*y;
+        m.m03 = 0.;
+        m.m10 = 2*x*y - 2*s*z;
+        m.m11 = s*s + y*y - x*x - z*z;
+        m.m12 = 2*y*z + 2*s*x;
+        m.m13 = 0.;
+        m.m20 = 2*x*z + 2*s*y;
+        m.m21 = 2*y*z - 2*s*x;
+        m.m22 = s*s + z*z - x*x - y*y; 
+        m.m23 = 0.;
+        // Extract translation information
+        var t = getTranslation();
+        m.m30 = t.x;
+        m.m31 = t.y;
+        m.m32 = t.z;
+        // pad last row
+        m.m33 = 1.;
+        return m;
+    }
+    
+    /**
+     * Linear interpolation between two dual quaternions, needs testing think it will work!?
+     * 
+     * @param a     The value at t = 0
+     * @param b     The value at t = 1
+     * @param t     A number in the range [0, 1]
+     * @return      The interpolated value
+     */
+    public static inline function lerp(a:DualQuaternion, b:DualQuaternion, t:Float):DualQuaternion
+    {
+        return new DualQuaternion( Quaternion.lerp(a.real, b.real,t)
+                                 , Quaternion.lerp(a.dual, b.dual,t)
+                                 );
+    }
+    
+    public static inline function slerp(a:DualQuaternion, b:DualQuaternion, t:Float):DualQuaternion
+    {
+        return new DualQuaternion( Quaternion.slerp(a.real, b.real,t)
+                                 , Quaternion.slerp(a.dual, b.dual,t)
+                                 );
+    }
+    
+    /**
+     * Sets real fields of this structure without allocation.
+     * does not normalize see 'create'
+     * 
+     * @param s
+     * @param x
+     * @param y
+     * @param z
+     * @return self
+     */
+    public inline function setReal(s:Float, x:Float, y:Float, z:Float):DualQuaternion
```

Comment:
Not sure these helpers are useful. The fields are forwarded, so probably easier to set it that way.

Reply:
removed